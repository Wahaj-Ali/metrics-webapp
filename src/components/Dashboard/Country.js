import React from 'react';
import PropTypes from 'prop-types';
import { Col } from 'react-bootstrap';
import styles from './country.module.css';

const Country = (props) => {
  const { country, detailed } = props;
  const { name, population, flags } = country;
  return (

    <Col className={styles.card}>
      <div className="flex-center column country">

        <img className={styles.flag} src={flags.svg} alt={`${name}Country Images`} />

        <div className={styles.descp}>
          <h4>{name}</h4>
          <p>
            {detailed && 'Poplution: '}
            {population}
          </p>
        </div>

      </div>
    </Col>

  );
};

Country.propTypes = {
  name: PropTypes.string,
  population: PropTypes.number,
  flags: PropTypes.any,
}.isRequired;

export default Country;
