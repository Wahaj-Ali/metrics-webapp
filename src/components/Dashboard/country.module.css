@import url('https://fonts.googleapis.com/css2?family=Montserrat&display=swap');

.card {
  margin: 10px;
  background: rgb(215, 55, 111);
  padding: 10px;
  width: 300px;
  height: 215px;
  display: flex;
  align-items: center;
  justify-content: center;
  box-shadow: 0 0 11px 8px #6c262c;
}

.list:nth-child(even) .card {
  background-color: rgb(161 35 79);
}

ul,
li {
  list-style: none;
  color: white;
}

.list {
  margin: -15px 0;
}

.descp {
  margin: 10px;
  padding: 5px;
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  position: relative;
  left: 20px;
}

.descp h4,
p {
  color: #fff;
}

.descp p {
  font-family: 'Montserrat', sans-serif;
}

.descpDiv {
  display: flex;
  flex-direction: column;
}

.countryContainer {
  display: flex;
}

.countrylist {
  color: white;
  font-size: 17px;
  display: flex;
  flex-flow: wrap;
  margin: 5px 40px;
  justify-content: center;
}

.flag {
  width: 150px;
  align-self: flex-start;
  display: flex;
  border-radius: 10px;
}

.direct {
  font-size: x-large;
  position: relative;
  top: 45px;
  left: 135px;
  width: 22px;
}

.singleCountry {
  color: white;
  text-decoration: none;
}

.statsheading {
  display: flex;
  margin: 5px 50px;
  color: #fff;
  text-transform: uppercase;
  font-size: 17px;
}

@media (max-width: 768px) {
  .countryContainer {
    display: inline-flex;
  }

  .countrylist {
    display: grid;
    grid-template-columns: 0.5fr 0.75fr;
    margin: 1px;
  }

  .card {
    width: 150px;
    margin: 3px;
  }

  .direct {
    top: 38px;
    left: 65px;
  }

  .searchBar {
    margin: 10px;
  }

  .statsheading {
    margin: 8px;
    font-size: 16px;
  }

  .descp p {
    font-size: 11px;
  }

  .flag {
    width: 100px;
  }
}
