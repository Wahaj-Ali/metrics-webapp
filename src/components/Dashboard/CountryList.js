import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { BsArrowRightCircle } from 'react-icons/bs';
import ReactPaginate from 'react-paginate';
import { fetchApi } from '../../Redux/pollution/pollutionSlice';
import Country from './Country';
import styles from './country.module.css';

const CountryList = () => {
  const countryList = useSelector((state) => state.pollutionSlice.countryList);

  const dispatch = useDispatch();

  const [search, setSearch] = useState('');

  const [currentPage, setCurrentPage] = useState(0);

  const isMobile = window.innerWidth <= 768;
  const itemsPerPage = isMobile ? 4 : 8;

  const handleChange = (e) => {
    setSearch(e.target.value.toLowerCase());
    setCurrentPage(0);
  };

  useEffect(() => {
    if (!countryList) {
      dispatch(fetchApi());
    }
  }, [dispatch, countryList]);

  // Handle page change
  const handlePageChange = (selected) => {
    setCurrentPage(selected.selected);
  };

  // Filter and paginate the data
  const filteredItems = countryList
    ? countryList.filter((filtered) => {
      if (search === '') {
        return filtered;
      }
      return filtered.name.toLowerCase().includes(search);
    })
    : [];

  const pageCount = Math.ceil(filteredItems.length / itemsPerPage);
  const displayedItems = filteredItems.slice(
    currentPage * itemsPerPage,
    (currentPage + 1) * itemsPerPage,
  );

  return (
    <>
      <h2 className={styles.statsheading}>Stats by Country Name</h2>

      <input
        className="search-input"
        type="text"
        value={search}
        aria-label="search"
        onChange={handleChange}
        placeholder="Search a country..."
      />
      <div className={styles.countryContainer}>

        <ul className={styles.countrylist}>
          {displayedItems.length > 0 ? (
            displayedItems.map((country) => (
              <li key={country.numericCode} className={styles.list}>
                <Link to={`/${country.numericCode}`} className={styles.singleCountry}>
                  <BsArrowRightCircle className={styles.direct} />
                  <Country country={country} detailed />
                </Link>
              </li>
            ))
          ) : (
            <p className="max-width flex-center">No matching countries found.</p>
          )}
        </ul>
        {/* Pagination component */}
        {countryList && (
          <ReactPaginate
            previousLabel="<"
            nextLabel=">"
            breakLabel=""
            pageCount={pageCount}
            pageRangeDisplayed={3}
            marginPagesDisplayed={0}
            onPageChange={handlePageChange}
            className={styles.pagination}
            pageClassName={styles.pageItem}
            pageLinkClassName={styles.pageLink}
            previousClassName={styles.pageItem}
            previousLinkClassName={styles.pageLink}
            nextClassName={styles.pageItem}
            nextLinkClassName={styles.pageLink}
            activeClassName={styles.active} //
          />
        )}
      </div>
    </>
  );
};

export default CountryList;
