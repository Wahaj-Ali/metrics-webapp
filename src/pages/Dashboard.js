import React from 'react';
import Country from '../components/Country';
import Header from '../components/Header';

const Dashboard = () => (
  <div>
    <Header />
    <Country />
  </div>
);

export default Dashboard;
