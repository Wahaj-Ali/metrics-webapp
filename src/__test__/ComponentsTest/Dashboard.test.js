import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import '@testing-library/jest-dom';
import Dashboard from '../../pages/Dashboard';
import store from '../../Redux/configureStore.js';


describe('Dashboard', () => {
    test('renders correctly', () => {
      const tree = renderer.create(
        <Provider store={store}>
        <Router>
          <Dashboard />
        </Router>,
        </Provider>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });