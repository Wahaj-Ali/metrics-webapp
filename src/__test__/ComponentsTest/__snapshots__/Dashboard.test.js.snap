// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard renders correctly 1`] = `
Array [
  <div>
    <div
      className="navbar"
      data-testid="header"
    >
      <div
        className="brand"
      >
        <h1>
          Population Index
        </h1>
      </div>
      <div
        className="settings row"
      >
        <div
          className="audio"
        >
          <svg
            fill="currentColor"
            height={25}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 1024 1024"
            width={25}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M512 624c93.9 0 170-75.2 170-168V232c0-92.8-76.1-168-170-168s-170 75.2-170 168v224c0 92.8 76.1 168 170 168zm330-170c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8 0 140.3-113.7 254-254 254S258 594.3 258 454c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8 0 168.7 126.6 307.9 290 327.6V884H326.7c-13.7 0-24.7 14.3-24.7 32v36c0 4.4 2.8 8 6.2 8h407.6c3.4 0 6.2-3.6 6.2-8v-36c0-17.7-11-32-24.7-32H548V782.1c165.3-18 294-158 294-328.1z"
            />
          </svg>
        </div>
        <div>
          <svg
            fill="currentColor"
            height={25}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width={25}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M413.967 276.8c1.06-6.235 1.06-13.518 1.06-20.8s-1.06-13.518-1.06-20.8l44.667-34.318c4.26-3.118 5.319-8.317 2.13-13.518L418.215 115.6c-2.129-4.164-8.507-6.235-12.767-4.164l-53.186 20.801c-10.638-8.318-23.394-15.601-36.16-20.801l-7.448-55.117c-1.06-4.154-5.319-8.318-10.638-8.318h-85.098c-5.318 0-9.577 4.164-10.637 8.318l-8.508 55.117c-12.767 5.2-24.464 12.482-36.171 20.801l-53.186-20.801c-5.319-2.071-10.638 0-12.767 4.164L49.1 187.365c-2.119 4.153-1.061 10.399 2.129 13.518L96.97 235.2c0 7.282-1.06 13.518-1.06 20.8s1.06 13.518 1.06 20.8l-44.668 34.318c-4.26 3.118-5.318 8.317-2.13 13.518L92.721 396.4c2.13 4.164 8.508 6.235 12.767 4.164l53.187-20.801c10.637 8.318 23.394 15.601 36.16 20.801l8.508 55.117c1.069 5.2 5.318 8.318 10.637 8.318h85.098c5.319 0 9.578-4.164 10.638-8.318l8.518-55.117c12.757-5.2 24.464-12.482 36.16-20.801l53.187 20.801c5.318 2.071 10.637 0 12.767-4.164l42.549-71.765c2.129-4.153 1.06-10.399-2.13-13.518l-46.8-34.317zm-158.499 52c-41.489 0-74.46-32.235-74.46-72.8s32.971-72.8 74.46-72.8 74.461 32.235 74.461 72.8-32.972 72.8-74.461 72.8z"
            />
          </svg>
        </div>
      </div>
    </div>
    <h2
      className="statsheading"
    >
      Stats by Country Name
    </h2>
    <div
      className="countryContainer"
    >
      <ul
        className="countrylist"
      >
        <p
          className="max-width flex-center"
        >
          Loading...
        </p>
      </ul>
    </div>
  </div>,
  ",",
]
`;
